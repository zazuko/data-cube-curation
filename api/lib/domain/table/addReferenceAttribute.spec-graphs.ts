import { turtle } from '@tpluscode/rdf-string'
import { schema } from '@tpluscode/rdf-ns-builders'
import { parseGraph } from '../../__tests-helpers__'
import { dataCube } from '../../namespaces'

const sourceWithColumn = (sourceId: string) => turtle`
    <${sourceId}> ${dataCube.column} <${sourceId}/column1> .
    <${sourceId}/column1> a ${dataCube.Column}; ${schema.name} "Column 1" .
`

export const sourceWithColumnGraph = (sourceId: string) => parseGraph(sourceWithColumn(sourceId))()
