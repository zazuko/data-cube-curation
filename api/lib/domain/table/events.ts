export interface TableEvents {
  FactTableCreated: {
    projectId: string
    sourceId: string
  }
}
