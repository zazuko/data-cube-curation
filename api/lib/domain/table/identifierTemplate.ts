import { parse } from 'uri-template'
import cf from 'clownface'
import { Dataset, NamedNode } from 'rdf-js'
import { getSourceColumns } from '../../read-graphs/source/getSourceColumns'
import { rdf, dataCube, schema } from '../../namespaces'

interface IdentifierColumn {
  id: string;
  name: string;
}

export async function extractColumns (sourceIdOrDataset: string | Dataset, template: string | null): Promise<IdentifierColumn[] | Error> {
  let columnNames: string[]
  try {
    columnNames = parse(template).expressions.map(ex => ex.params[0].name).map(decodeURIComponent)
  } catch (e) {
    return e
  }

  const dataset = typeof sourceIdOrDataset === 'string'
    ? await getSourceColumns(sourceIdOrDataset)
    : sourceIdOrDataset

  const columnsNotFound: string[] = []
  const columns = cf({ dataset })
    .has(rdf.type, dataCube.Column)
  const columnIds = columnNames.reduce((identifiers, name) => {
    const columnNameTerm = columns.has(schema('name'), name).term as NamedNode
    if (!columnNameTerm) {
      columnsNotFound.push(name)
      return identifiers
    }

    return [
      ...identifiers,
      {
        id: columnNameTerm.value,
        name,
      }]
  }, [] as IdentifierColumn[])

  if (columnsNotFound.length > 0) {
    return new Error(`Columns ${columnsNotFound.join(', ')} were not found in source`)
  }

  return columnIds
}
