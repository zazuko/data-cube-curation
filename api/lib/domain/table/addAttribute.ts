import { factory, DomainError } from '@tpluscode/fun-ddr'
import { Table } from './index'
import { Attribute } from '../attribute'
import { AttributeEvents } from '../attribute/events'
import { expand } from '@zazuko/rdf-vocabularies'
import { existsInTableSource } from '../../read-graphs/table'
import { errorFactory } from '../error-helper'

interface AddAttributeCommand {
  name: string;
  columnId: string;
  predicate: string;
  datatype?: string;
  language?: string;
}

export const addAttribute = factory<Table, AddAttributeCommand, Attribute>(async (table, command, emitter) => {
  const error = errorFactory(table, 'Cannot add attribute to table')

  if (!command.name) {
    throw error('Name missing')
  }
  if (!command.predicate) {
    throw error('Predicate missing')
  }
  if (!command.columnId) {
    throw error('Column missing')
  }
  if (command.datatype && command.language) {
    throw error('Datatype and language cannot be used together')
  }
  if (!await existsInTableSource(table['@id'], command.columnId)) {
    throw error("Column not found or it does not belong to the table's source")
  }

  const attributeId = `${table['@id']}/attribute/${encodeURIComponent(command.name)}`

  emitter.emit<AttributeEvents, 'AttributeAdded'>('AttributeAdded', {
    name: command.name,
    tableId: table['@id'],
    columnId: command.columnId,
    predicate: command.predicate,
    datatype: command.datatype,
    language: command.language,
  })

  return {
    '@id': attributeId,
    '@type': 'Attribute',
    tableId: table['@id'],
    name: command.name,
    column: command.columnId,
    predicate: command.predicate,
    datatype: command.datatype || expand('xsd:string'),
    language: command.language,
  }
})
