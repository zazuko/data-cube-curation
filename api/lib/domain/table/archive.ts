import { mutate } from '@tpluscode/fun-ddr'
import { Table } from './index'
import { TableEvents } from './events'
import { ask } from '../../sparql'
import { dataCube } from '../../namespaces'
import { getClient } from '../../read-graphs/sparqlClient'

export const archiveTable = mutate<Table, never, TableEvents>(async (state, cmd, emitter) => {
  const isFactTable = await ask(`<${state['@id']}> a dataCube:FactTable`)
    .prefixes({ dataCube })
    .execute(getClient())

  emitter.emit.TableArchived({
    isFactTable,
    projectId: state.projectId,
  })

  return state
})
