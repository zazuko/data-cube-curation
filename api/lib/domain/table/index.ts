import { initialize, Entity } from '@tpluscode/fun-ddr'
import urlSlug from 'url-slug'
import { TableEvents } from './events'

export { update } from './update'

export interface Table extends Entity {
  sourceId: string;
  projectId: string;
  tableName: string;
}

export interface FactTable extends Table {
  identifierTemplate: string | null;
}

export interface DimensionTable extends Table {
  identifierTemplate: string;
}

interface CreateTableCommand {
  sourceId: string;
  projectId: string;
  tableName: string;
  identifierTemplate: string | null;
}

export const createTable = initialize<Table, CreateTableCommand, TableEvents>((cmd, emitter) => {
  const tableId = `${cmd.projectId}/table/${urlSlug(cmd.tableName)}`

  emitter.emit.FactTableCreated({
    projectId: cmd.projectId,
    sourceId: cmd.sourceId,
    tableName: cmd.tableName,
    identifierTemplate: cmd.identifierTemplate,
  })

  return {
    '@id': tableId,
    '@type': 'Table',
    projectId: cmd.projectId,
    sourceId: cmd.sourceId,
    tableName: cmd.tableName,
    identifierTemplate: cmd.identifierTemplate,
  }
})
