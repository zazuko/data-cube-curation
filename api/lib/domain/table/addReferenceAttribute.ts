import uuid from 'uuid'
import cf from 'clownface'
import { NamedNode } from 'rdf-js'
import { factory } from '@tpluscode/fun-ddr'
import { Table } from './index'
import { ReferenceAttribute } from '../attribute'
import { errorFactory } from '../error-helper'
import { AttributeEvents } from '../attribute/events'
import { getSourceColumns } from '../../read-graphs/source/getSourceColumns'
import { getTableSourceId } from '../../read-graphs/table'
import { dataCube, rdf } from '../../namespaces'
import { extractColumns } from './identifierTemplate'
import { getIdentifierTemplate } from '../../read-graphs/table/dimensionTable'

interface ColumnMapping {
  sourceColumnId: string;
  referencedColumnId: string;
}

interface AddReferenceAttributeCommand {
  predicate: string;
  referencedTableId: string;
  columnMappings: ColumnMapping[];
}

function validateColumns (columnsDataset: any, columnMappings: ColumnMapping[]) {
  return async function (selectColumn: (m: ColumnMapping) => string) {
    const errors: string[] = []
    const columns = cf({ dataset: columnsDataset })
      .has(rdf.type, dataCube.Column)
      .terms
      .map((term: NamedNode) => term.value)

    columnMappings.forEach(mapping => {
      const columnId = selectColumn(mapping)

      if (!columns.includes(columnId)) {
        errors.push(`Column ${columnId} not found`)
      }
    })

    return errors
  }
}

async function validateSourceColumns (columnsDataset: any, columnMappings: ColumnMapping[]) {
  return validateColumns(columnsDataset, columnMappings)(mapping => mapping.sourceColumnId)
}

async function validateReferencedColumns (tableId: string, columnsDataset: any, columnMappings: ColumnMapping[]) {
  const errors = await validateColumns(columnsDataset, columnMappings)(mapping => mapping.referencedColumnId)

  const template = await getIdentifierTemplate(tableId)
  const identifierColumns = await extractColumns(columnsDataset, template)
  if (identifierColumns instanceof Error) throw identifierColumns

  identifierColumns.forEach(column => {
    if (!columnMappings.find(mapping => mapping.referencedColumnId === column.id)) {
      errors.push(`Column ${column} is part of referenced table's identifier but not used in the attribute`)
    }
  })

  return errors
}

export const addReferenceAttribute = factory<Table, AddReferenceAttributeCommand, ReferenceAttribute>(async (table, command, emitter) => {
  const DomainError = errorFactory(table, 'Cannot add attribute to table')
  if (!command.predicate) {
    throw new DomainError('Predicate missing')
  }

  if (!command.referencedTableId) {
    throw new DomainError('Referenced table missing')
  }

  const referencedSourceId = await getTableSourceId(command.referencedTableId)
  if (!referencedSourceId) {
    throw new DomainError('Referenced table not found')
  }

  const sourceTableColumns = await getSourceColumns(table.sourceId)
  const referencedTableColumns = await getSourceColumns(referencedSourceId)

  const attributeId = `${table['@id']}/attribute/${uuid()}`
  const sourceColumnErrors = await validateSourceColumns(sourceTableColumns, command.columnMappings)
  const referencedColumnsErrors = await validateReferencedColumns(command.referencedTableId, referencedTableColumns, command.columnMappings)
  if (sourceColumnErrors.length > 0 || referencedColumnsErrors.length > 0) {
    throw new DomainError(...sourceColumnErrors, ...referencedColumnsErrors)
  }

  emitter.emit<AttributeEvents, 'ReferenceAttributeAdded'>('ReferenceAttributeAdded', {
    tableId: table['@id'],
    predicate: command.predicate,
    referencedTableId: command.referencedTableId,
    columnMappings: command.columnMappings,
  })

  return {
    '@id': attributeId,
    '@type': ['Attribute', 'ReferenceAttribute'],
    tableId: table['@id'],
    predicate: command.predicate,
    referencedTableId: command.referencedTableId,
    columnMappings: command.columnMappings,
  }
})
