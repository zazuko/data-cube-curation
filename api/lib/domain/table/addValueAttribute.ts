import { factory } from '@tpluscode/fun-ddr'
import uuid from 'uuid'
import { Table } from './index'
import { ValueAttribute } from '../attribute'
import { AttributeEvents } from '../attribute/events'
import { expand } from '@zazuko/rdf-vocabularies'
import { existsInTableSource } from '../../read-graphs/table'
import { errorFactory } from '../error-helper'

interface AddValueAttributeCommand {
  columnId?: string;
  propertyTemplate?: string;
  datatype?: string;
  language?: string;
}

export const addValueAttribute = factory<Table, AddValueAttributeCommand, ValueAttribute>(async (table, command, emitter) => {
  const DomainError = errorFactory(table, 'Cannot add attribute to table')

  if (!command.propertyTemplate) {
    throw new DomainError('Property template missing')
  }
  if (!command.columnId) {
    throw new DomainError('Column missing')
  }
  if (command.datatype && command.language) {
    throw new DomainError('Datatype and language cannot be used together')
  }
  if (!await existsInTableSource(table['@id'], command.columnId)) {
    throw new DomainError("Column not found or it does not belong to the table's source")
  }

  const attributeId = `${table['@id']}/attribute/${uuid()}`

  emitter.emit<AttributeEvents, 'ValueAttributeAdded'>('ValueAttributeAdded', {
    tableId: table['@id'],
    columnId: command.columnId,
    propertyTemplate: command.propertyTemplate,
    datatype: command.datatype || expand('xsd:string'),
    language: command.language,
  })

  return {
    '@id': attributeId,
    '@type': ['Attribute', 'ValueAttribute'],
    tableId: table['@id'],
    column: command.columnId,
    propertyTemplate: command.propertyTemplate,
    datatype: command.datatype || expand('xsd:string'),
    language: command.language,
  }
})
