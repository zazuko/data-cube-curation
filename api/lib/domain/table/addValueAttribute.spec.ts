import { addValueAttribute, AddValueAttributeCommand } from './addValueAttribute'
import { Table } from './index'
import { expand } from '@zazuko/rdf-vocabularies'
import { existsInTableSource } from '../../read-graphs/table'
import { xsd } from '../../namespaces'

jest.mock('../../read-graphs/table')

const existsInTableSourceMock = existsInTableSource as jest.Mock

describe('table', () => {
  let table: Table

  const command: Partial<AddValueAttributeCommand> = {
    columnId: 'source/column',
    propertyTemplate: expand('schema:name'),
  }

  beforeEach(() => {
    table = {
      '@id': 'example/table',
      '@type': 'Table',
      sourceId: 'unimportant/source',
      projectId: 'unimportant/project',
      tableName: 'Example',
    }
  })

  describe('add value attribute', () => {
    it('errors when column is missing', async () => {
      // given
      const cmd = { ...command }
      delete cmd.columnId

      // when
      const result = await addValueAttribute(table, cmd)

      // then
      await expect(result.error).resolves.toBeInstanceOf(Error)
    })

    it('errors when predicate is missing', async () => {
      // given
      const cmd = { ...command }
      delete cmd.propertyTemplate

      // when
      const result = await addValueAttribute(table, cmd)

      // then
      await expect(result.error).resolves.toBeInstanceOf(Error)
    })

    it('errors when datatype is used together with language', async () => {
      // given
      const cmd = {
        ...command,
        datatype: expand('xsd:integer'),
        language: 'fr',
      }

      // when
      const result = await addValueAttribute(table, cmd)

      // then
      await expect(result.error).resolves.toBeInstanceOf(Error)
    })

    it('return expected state', async () => {
      // given
      existsInTableSourceMock.mockResolvedValueOnce(true)

      // when
      const result = await addValueAttribute(table, command)

      // then
      const state = await result.state
      expect(state).toMatchSnapshot({
        '@id': expect.stringMatching(new RegExp('example/table/attribute/.+')),
      })
    })

    it('emits an event', async () => {
      // given
      existsInTableSourceMock.mockResolvedValueOnce(true)

      // when
      const result = await addValueAttribute(table, command)

      // then
      await expect(result.events).resolves.toMatchSnapshot()
    })

    describe('with datatype parameters', () => {
      it('persists the parameters', async () => {
        // given
        existsInTableSourceMock.mockResolvedValueOnce(true)
        command.datatype = xsd.date.value
        command.parameters = {
          format: 'YYYY/mm/dd',
        }

        // when
        const result = await addValueAttribute(table, command)

        // then
        const state = await result.state
        expect(state).toMatchSnapshot({
          '@id': expect.stringMatching(new RegExp('example/table/attribute/.+')),
        })
      })

      it('emits an event', async () => {
        // given
        existsInTableSourceMock.mockResolvedValueOnce(true)
        command.datatype = xsd.date.value
        command.parameters = {
          format: 'YYYY/mm/dd',
        }

        // when
        const result = await addValueAttribute(table, command)

        // then
        await expect(result.events).resolves.toMatchSnapshot()
      })
    })
  })
})
