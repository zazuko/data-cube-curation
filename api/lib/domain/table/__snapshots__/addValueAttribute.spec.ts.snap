// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table add value attribute emits an event 1`] = `
Array [
  Object {
    "data": Object {
      "columnId": "source/column",
      "datatype": "http://www.w3.org/2001/XMLSchema#string",
      "language": undefined,
      "predicate": "http://schema.org/name",
      "tableId": "example/table",
    },
    "name": "ValueAttributeAdded",
  },
]
`;

exports[`table add value attribute return expected state 1`] = `
Object {
  "@id": StringMatching /example\\\\/table\\\\/attribute\\\\/\\.\\+/,
  "@type": Array [
    "Attribute",
    "ValueAttribute",
  ],
  "column": "source/column",
  "datatype": "http://www.w3.org/2001/XMLSchema#string",
  "language": undefined,
  "predicate": "http://schema.org/name",
  "tableId": "example/table",
}
`;
